pallets/utxo/src/rewards.rs

```diff
@@ -0,0 +1,207 @@
+use sp_runtime::Percent;
+use sp_std::convert::TryInto;
+use sp_core::H256;
+
+/// handle event when a block author is found.
+impl<T> pallet_authorship::EventHandler<T::AccountId, T::BlockNumber> for Pallet<T>
+where
+    T: Config + pallet_authorship::Config,
+{
+    fn note_author(author: T::AccountId) {
+        match convert_to_h256::<T>(&author) {
+            Ok(author_h256) => {
+                // store the block author. Reward during the `fn finalize()` phase.
+                <BlockAuthor<T>>::put(author_h256);
+            }
+            Err(e) => {
+                log::warn!("failed to find author: {:?}", e);
+            }
+        }
+    }
+
+    fn note_uncle(_author: T::AccountId, _age: T::BlockNumber) {
+        log::info!("TODO: no support for this. Or is there...?");
+    }
+}
+
+
+/// checks at what period the given block number belongs to.
+fn get_block_period<T:Config>(block_number: T::BlockNumber) -> u8 {
+    let reduction_period: T::BlockNumber = T::RewardReductionPeriod::get();
+
+    // get at what period the current block number is at.
+    block_number.checked_div(&reduction_period)
+        .map_or(u8::MAX, |f|{
+            // convert data type from T::BlockNumber to u8
+            f.try_into().unwrap_or(u8::MAX)
+        })
+}
+
+
+/// Returns the newly reduced reward amount for a Block Author.
+/// How much a reward is reduced, will be based on the config's`RewardReductionFraction`.
+fn get_block_author_reward<T: Config>(block_number: T::BlockNumber) -> Value {
+    let reward_amount = T::InitialReward::get();
+    let reduction_period: T::BlockNumber = T::RewardReductionPeriod::get();
+    let reduction_fraction = T::RewardReductionFraction::get().deconstruct();
+
+    // no computation needed; current block number is still within the reduction period.
+    if block_number < reduction_period {
+        return reward_amount;
+    }
+
+    // The maximum period before approaching to 100% reduction
+    let max_period = (100u8 / reduction_fraction) - 1;
+
+    // current period of the given block number
+    let current_period = get_block_period::<T>(block_number);
+
+    // when current_period has reached or exceeded the maximum period of reduction,
+    // return the least amount to reward.
+    if current_period > max_period {
+        // TODO: as a note, this is only testnet specific to reward at least 1
+        return 1;
+    }
+
+    // compute for the updated reduction % , given the current period.
+    let updated_reduction_fraction = Percent::from_percent(reduction_fraction * current_period );
+
+    reward_amount - updated_reduction_fraction.mul_ceil(reward_amount)
+}
+
+fn insert_to_utxo_store<T:Config>(block_number:T::BlockNumber, block_author:H256, reward:Value) {
+    let utxo = TransactionOutput::new_pubkey(reward, block_author.clone());
+
+    let hash = {
+        let b_num = block_number.saturated_into::<u64>();
+        BlakeTwo256::hash_of(&(&utxo, b_num, "author_reward"))
```

Comment:
This is a very bad way of doing it, but let's let it pass for the testnet as I kind of see why you're doing it this way. Please, please oh, please make a note somewhere that this has to be changed. It seems extremely arbitrary.

There has to be a well-defined way to find the block reward, not an arbitrary emergent property from something that's being done to solve a technical issue.

Later, we'll change this to something like coinstake, where the previous locked stake is spent and is replaced with a UTXO that has the previous amount + reward.